### PR TITLE
Pyn sale split quote2many order

### PR DESCRIPTION
**Pull request by phongyanon**
please review my code.

### CODE REVIEW

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    use_merge = fields.Boolean(
+        string='Merge Order Line',
+        states={'done': [('readonly', True)]},
+        default=False,
+        help="Merge order line to be single line.",
+    )
+    amount_untaxed_order_plan  = fields.Float(
+        string='Amount Untaxed',
+        readonly=True,
+        related='amount_untaxed',
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py]: This related looks suspecious. Why do we need another field?

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    @api.onchange('order_plan_ids')
+    def _compute_customer_amount(self):
+        for order in self:
+            self.customer_amount = len(self.order_plan_ids)
+
+    @api.multi
+    def merge_sale_order_line(self, amount_price):
+        self.ensure_one()
+        order_lines = self.order_line
+        description = (self.name or '') + ' ' + \
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py]: Use string injection better, there are 2 ways,
```
description = '%s %s %s' % ('A', 'B', 'B')
description = '%(a)s %(b)s %(b)s' % {'a': 'A', 'b': 'B'}
```

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+            'order_lines_group': 'before',
+            'sale_layout_custom_group_id': None,
+            'section_code': None,
+            'product_id': None,
+            'price_unit': amount_price /
+                          (1.0 - (order_lines[0].discount / 100.0)),
+            'product_uom_qty': 1.0,
+            'name': description,
+        })
+        for line in order_lines:
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py]: You can use `order_lines.unlink()`
Even here is no line in order_lines, it is ok.

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+        return res
+
+    @api.multi
+    def adjust_price_sale_order_line(self, amount_price):
+        self.ensure_one()
+        order_line = self.order_plan_ids
+        if order_line and self.sale_order_mode:
+            if self.sale_order_mode == 'change_quantity':
+                for line in self.order_line:
+                    new_quantity = amount_price / self.amount_untaxed
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py]: Check for division by zero. Also, always do float_round() if there is float division.

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+        self._amount_all()
+
+    @api.multi
+    def validate_sale_order_plan(self):
+        self.ensure_one()
+        order_plan = self.order_plan_ids
+        order_plan_amount = sum(order_plan.mapped('sale_order_amount'))
+        order_plan_percent = sum(order_plan.mapped('sale_order_percent'))
+        if (self.use_merge is False) and (self.sale_order_mode is False):
+            raise Warning(_("Should select sale order mode \
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py]: You can use,
```
raise Warning(_("Should select sale order mode "
                "if not use merge sale order line"))
```

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+            raise Warning(_("Order plan percent must more than zero"))
+        return order_plan
+
+    @api.multi
+    def action_validate_before_convert2order(self):
+        for order in self:
+            if self.use_multi_customer and self.order_plan_ids:
+                order_plan = self.validate_sale_order_plan()
+                ctx = self._context.copy()
+                current_date = datetime.date.today()
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py]: Use `fields.Date.context_today(self)`

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+                            plan.sale_order_amount)
+                    plan.write({'order_ref_id': new_sale_order.id})
+                order.signal_workflow('convert_to_order')
+            else:
+                self.action_button_convert_to_order()
+
+    @api.multi
+    def action_view_sale_order(self):
+        self.ensure_one()
+        domain = [
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py]: Use existing action if already exists, i.e.,
```
action = self.env.ref('external_id_of_action')
res = actions.read()[0]
```

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+            'views': [[False, "tree"], [False, "form"]],
+            'domain': domain,
+            'context': "{'active': True}"
+        }
+
+    @api.multi
+    def action_cancel_draft_sale_orders(self):
+        self.ensure_one()
+        sale_order_ids = self.env['sale.order'].search([
+            '&', ('quote_id', 'like', self.id),
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py:178]: When write domain. If it is "and", you don't need '&'
But no blocking.

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    @api.onchange('sale_order_percent')
+    def _onchange_sale_order_percent(self):
+        self.ensure_one()
+        precision = self.env['decimal.precision']
+        prec = precision.precision_get('Account')
+        subtotal = self.quote_id.amount_untaxed
+        self.sale_order_amount = round(self.sale_order_percent / 100.0 *
+                                       subtotal or 0.0, prec)
+
+    @api.multi
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py]: For @api.onchange() we don't need @api.multi()

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+    order_ref_state = fields.Selection(
+        related='order_ref_id.state',
+        string='State',
+    )
+    order_ref_name = fields.Char(
+        related='order_ref_id.name',
+        string='Name',
+    )
+
+    @api.multi
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py]: Don't need @api.multi

File: specific/cmo_sale_split_quote2many_order/models/sale.py [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+    @api.onchange('sale_order_percent')
+    def _onchange_sale_order_percent(self):
+        self.ensure_one()
```
Comment (kittiu) [inline specific/cmo_sale_split_quote2many_order/models/sale.py]: For onchange, the "self" is not a record yet. 
No need self.ensure_one()

**Comment by kittiu** [comment]
Thanks!

File: specific/cmo_sale_split_quote2many_order/models/sale.py (hunk 4 above)
Comment (phongyanon) [reply specific/cmo_sale_split_quote2many_order/models/sale.py]: `from openerp.tools.float_utils import float_round as round`
i use this code for replace round as a float round

File: specific/cmo_sale_split_quote2many_order/models/sale.py (hunk 1 above)
Comment (phongyanon) [reply specific/cmo_sale_split_quote2many_order/models/sale.py]: i show amount_untaxed fields 2 places in views.

Comment (kittiu) [reply specific/cmo_sale_split_quote2many_order/models/sale.py]: OK

File: specific/cmo_sale_split_quote2many_order/models/sale.py (hunk 4 above)
Comment (kittiu) [reply specific/cmo_sale_split_quote2many_order/models/sale.py]: OK

**Pull request merged**